### ISSUE TITLE
Should lua script be reloaded into cache on NOSCRIPT failure?

### ISSUE THREAD

**Issue by ebertz** [luin/ioredis#1438]
https://github.com/luin/ioredis/blob/62b6a648910eccc3d83a3acd2db873704fd2080a/lib/script.ts#L35-L51

When a Lua script is loaded using defineCommand(), ioredis will load the script into cache memory, and attempt to reference it using the SHA.

If the redis cluster is restarted or otherwise flushes the script cache, calling the custom script will first attempt to do an EVALSHA, then upon receiving a NOSCRIPT error, will call EVAL with the lua code stored in app memory.

Unless the application adds logic to reload the script, each subsequent call will always do 2 operations: a failed EVALSHA followed by an EVAL. This will obviously affect performance.

Instead of doing EVALSHA followed by EVAL each time the missing script is called, could ioredis instead handle NOSCRIPT errors by reloading the lua script into the redis script cache? That way, subsequent calls will be successful when trying the initial EVALSHA

**Comment by marcbachmann** [comment]
In the pipeline code scripts are handled differently: https://github.com/luin/ioredis/blob/master/lib/pipeline.ts#L376
They get loaded once and not retried unless a client side interval expires.

So in my case `EVAL` doesn't even get executed.

The following workaround will reliably work if `maxScriptsCachingTime` is set to a higher value, as it will force a script load on reconnect:
```js
const redis = new Redis({
  maxScriptsCachingTime: 3600
})

redis.on('reconnecting', () => {
  redisClient._addedScriptHashes = {}
  log.debug({instance}, `Redis connection reconnecting`)
})
```

I guess we can just put that logic into the disconnect function: https://github.com/luin/ioredis/blob/b8177479c348aa4bbd467fa944d61fe9b35aec19/lib/redis/index.ts#L438-L441

```patch
 Redis.prototype.disconnect = function (reconnect) {
   clearInterval(this._addedScriptHashesCleanInterval);
   this._addedScriptHashesCleanInterval = null;
+  this._addedScriptHashes = {};
```

**Comment by marcbachmann** [comment]
Fixed in #1499
An EVAL always loads a script automatically into the cache. With the update, ioredis now tries to send evalsha on a second execution. This will result in much less network traffic if the scripts have are bigger.

**Comment by marcbachmann** [comment]
The improvements landed on `main` and [5.0.0-beta.1](https://www.npmjs.com/package/ioredis/v/5.0.0-beta.1)
